src/inabox/inabox-position-observer.js

```diff
@@ -0,0 +1,146 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../src/services';
+import {registerServiceBuilderForDoc} from '../../src/service';
+import {
+  moveLayoutRect,
+} from '../../src/layout-rect';
+import {
+  serializeMessage,
+  MessageType,
+} from '../../src/3p-frame-messaging';
+import {tryParseJson} from '../../src/json.js';
+import {Observable} from '../../src/observable';
+import {
+  PosObViewportInfoDef,
+} from '../service/position-observer/position-observer-viewport-info';
+import {
+  PositionObserver,
+} from '../service/position-observer/position-observer-impl';
+
+/**
+ * @implements {PosObViewportInfoDef}
+ */
+class PosObViewportInfoInabox {
+  /**
+   * @param {!../service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    this.ampdoc = ampdoc;
+    this.win_ = ampdoc.win;
+    this.onMessageReceivedObservers_ = new Observable();
+    this.boundOnMessageEventListener_ =
+        event => this.onMessageReceivedObservers_.fire(event);
+    this.viewportBox_ = null;
+    this.iframePosition_ = null;
+    this.sentinel = null;
+  }
+
+  connect() {
+    this.win_.addEventListener('message', this.boundOnMessageEventListener_);
+    const object = {};
+    const dataObject = tryParseJson(this.win_.name);
+    if (dataObject) {
+      this.sentinel = dataObject['_context']['sentinel'];
+    }
+    object.type = MessageType.SEND_POSITIONS_HIGH_FIDELITY;
+    this.win_.parent./*OK*/postMessage(serializeMessage(
+        MessageType.SEND_POSITIONS_HIGH_FIDELITY, this.sentinel),
+        '*'
+    );
+  }
+
+  disconnect() {
+    this.win_.removeEventListener('message', this.boundOnMessageEventListener_);
+  }
+
+  /**
+   * @param {function()} unusedCallback
+   * @return {function()}
+   */
+  onScroll(unusedCallback) {
+    return () => {};
+  }
+
+  /**
+   * @param {function(?)} unusedCallback
+   * @return {function()}
+   */
+  onResize(unusedCallback) {
+    return () => {};
+  }
+
+  /**
+   * @param {function(?)} callback
+   * @return {function()}
+   */
+  onHostMessage(callback) {
+    return this.onMessageReceivedObservers_.add(callback);
+  }
+
+  getSize() {
+    return this.viewportBox_;
+  }
+
+  /**
+   * @param {!Element} element
+   */
+  getLayoutRect(element) {
+    if (!this.iframePosition_) {
+      // If not receive iframe position from host, or if iframe is outside vp
+      return null;
+    }
+    if (!element.inIframePositionRect) {
```

Comment:
Yes, but not because `getLayoutRect` is called before. 
It happens when the old `inIframePositionRect` is outdated, and we clear the stored value.  For example, if the inabox iframe gets resized, we would want to refresh the `inIframePositionRect` value.